### PR TITLE
- rewrite CAST implementation

### PR DESCRIPTION
**Pull request by be-marc**
* data leakage of space and time variables in the old version.
* space and time variables are set with a column role now.
* sampling of folds was not random in the old version. The `create_folds_caret()` function draws a stratified sample on row ids. This should be a random sample.
* removes caret code

 This new implementation should be reviewed in depth.

### CODE REVIEW

**Review by pat-s — COMMENTED** [review]
The following test which compares against `CAST::CreateSpacetimeFolds` fails. 

```r
data_only = tsk("cookfarm")$data()
task = tsk("cookfarm")
task$set_col_roles("SOURCEID", roles = "space")
rsp = rsmp("sptcv_cstf", folds = 10)
set.seed(42)
rsp$instantiate(task)
rsp$test_set(1)

set.seed(42)
indices = CAST::CreateSpacetimeFolds(data_only,
                                     spacevar = "SOURCEID")

expect_equal(rsp$instance$train, indices$index)
expect_equal(rsp$instance$test, indices$indexOut)
```

I think keeping at least one test against the upstream implementation could be worth it, what do you think?

File: tests/testthat/test-ResamplingSptCVCstf.R [hunk 1]
```diff
@@ -1,61 +1,519 @@
...
+    "L1", 1, "A",
+    "L1", 2, "A",
+    "L1", 3, "A",
+    "L2", 1, "A",
+    "L2", 2, "A",
+    "L2", 3, "A",
+    "L3", 1, "B",
+    "L3", 2, "B",
+    "L3", 3, "B"
+  )
```
[the remark this answers is not in the document]
Comment (be-marc) [reply tests/testthat/test-ResamplingSptCVCstf.R:14]: Why should we?  It is much easier to understand what happens in the unit test if you can see the data. We should deactivate precommit for `rowwise_table`, it destroys the formatting.

[the remark this answers is not in the document]
Comment (pat-s) [reply tests/testthat/test-ResamplingSptCVCstf.R:14]: Just to remove redundant code, we have so many lines in various tests creating these tables?

You mean precommit invoking styler?

**Comment by pat-s** [comment]
> Currently, the test would fail because the upstream implementation draws a stratified sample and our implementation a random sample. I think it is much more important that our implementation matches the description of the algorithm in Meyer et al. (2018). We can guarantee that with our unit tests.

Well sure, what I meant was more on the general side: {mlr3spatiotempcv} aims to do the same as the upstream pkgs. If we find issues in the upstream implementation that should be fixed this is great, but in the long run we should have tests that verify the integrity against both the theoretical underpinning of the method and the upstream implementation, no? Having tests against both is even better than only testing against upstream and "hoping" it has been implemented correctly.

**Comment by be-marc** [comment]
> > Currently, the test would fail because the upstream implementation draws a stratified sample and our implementation a random sample. I think it is much more important that our implementation matches the description of the algorithm in Meyer et al. (2018). We can guarantee that with our unit tests.
> 
> Well sure, what I meant was more on the general side: {mlr3spatiotempcv} aims to do the same as the upstream pkgs. If we find issues in the upstream implementation that should be fixed this is great, but in the long run we should have tests that verify the integrity against both the theoretical underpinning of the method and the upstream implementation, no? Having tests against both is even better than only testing against upstream and "hoping" it has been implemented correctly.

Sure you can add the test against the upstream implementation again when it is fixed. As long as you don't add a CARET code to mlr3 again to make it work. Both versions draw a random sample and they could be different even if we set the same seed beforehand.

File: tests/testthat/test-ResamplingSptCVCstf.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (be-marc) [reply tests/testthat/test-ResamplingSptCVCstf.R:14]: I don't care about redundant code in unit tests. I want to understand the test quickly when there's a bug. 

Yes it removes the whitespaces https://github.com/mlr-org/mlr3spatiotempcv/pull/174/commits/b2a0ba85ea1c65a95784f3ce505a1dd000bc7b2c.

[the remark this answers is not in the document]
Comment (pat-s) [reply tests/testthat/test-ResamplingSptCVCstf.R:14]: styler respects whitespace between columns in such situation but they must align with the leading `~`.

E.g. the following is not changed

```r
data = rowwise_table(
  ~polygon, ~day, ~class,
  "L1",     1,    "A",
  "L1",     2,    "A",
  "L1",     3,    "A",
  "L2",     1,    "A",
  "L2",     2,    "A",
  "L2",     3,    "A",
  "L3",     1,    "B",
  "L3",     2,    "B",
  "L3",     3,    "B"
)
```

Would that work for you? In your example the content of the `day` column was too far too the right.

**Comment by pat-s** [comment]
> Sure you can add the test against the upstream implementation again when it is fixed. As long as you don't add CARET code to mlr3 again to make it work. Both versions draw a random sample and they could be different even if we set the same seed beforehand.

Another option would be to expect differences/errors when comparing with upstream. This way we would notice when upstream changes (i.e. match again with our implementation). Also by doing so we make clear how we currently testing things and why we are off compared to upstream (for users inspecting the code). What do you think?

> As long as you don't add CARET code to mlr3 

What is the reasoning here? and why mlr3?
I don't mind having CAST and caret in suggests for testing reasons. In fact, I think suggests can contain any number of pkgs if they are helpful for testing or other reasons. To me only Imports is important.

**Review by pat-s — APPROVED** [review]
@be-marc Checks are now passing fur CLUTO, we could merge from my POV.

I haven't checked the new CAST implementation yet though, this would need more time.

**Pull request merged**